// r1-create/dist/index.js
"use strict";
/**
 * R1 Create SDK
 * Unofficial community SDK for building R1/RabbitOS plugins
 *
 * Features:
 * - Hardware access (accelerometer, touch, PTT, scroll)
 * - Secure/plain storage with Base64 encoding
 * - LLM interaction and messaging
 * - Optimized UI utilities for 240x282px display
 * - Media APIs (camera, microphone, speaker)
 * - Event-driven architecture
 * - Mobile performance optimizations
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createR1App = exports.initializeR1 = exports.r1 = exports.R1Plugin = exports.R1SDK = exports.speaker = exports.microphone = exports.camera = exports.MediaUtils = exports.SpeakerAPI = exports.MicrophoneAPI = exports.CameraAPI = exports.ui = exports.R1UI = exports.R1_DIMENSIONS = exports.R1Component = exports.PerformanceUtils = exports.LayoutUtils = exports.DOMUtils = exports.CSSUtils = exports.llmHelpers = exports.messaging = exports.LLMHelpers = exports.R1Messaging = exports.storage = exports.R1Storage = exports.Base64Utils = exports.deviceControls = exports.DeviceControls = exports.hardwareEvents = exports.touch = exports.accelerometer = exports.HardwareEvents = exports.TouchAPI = exports.AccelerometerAPI = void 0;
// Type definitions
__exportStar(require("./types"), exports);
// Hardware APIs
var hardware_1 = require("./hardware");
Object.defineProperty(exports, "AccelerometerAPI", { enumerable: true, get: function () { return hardware_1.AccelerometerAPI; } });
Object.defineProperty(exports, "TouchAPI", { enumerable: true, get: function () { return hardware_1.TouchAPI; } });
Object.defineProperty(exports, "HardwareEvents", { enumerable: true, get: function () { return hardware_1.HardwareEvents; } });
Object.defineProperty(exports, "accelerometer", { enumerable: true, get: function () { return hardware_1.accelerometer; } });
Object.defineProperty(exports, "touch", { enumerable: true, get: function () { return hardware_1.touch; } });
Object.defineProperty(exports, "hardwareEvents", { enumerable: true, get: function () { return hardware_1.hardwareEvents; } });
// New device controls
Object.defineProperty(exports, "DeviceControls", { enumerable: true, get: function () { return hardware_1.DeviceControls; } });
Object.defineProperty(exports, "deviceControls", { enumerable: true, get: function () { return hardware_1.deviceControls; } });
// Storage APIs
var storage_1 = require("./storage");
Object.defineProperty(exports, "Base64Utils", { enumerable: true, get: function () { return storage_1.Base64Utils; } });
Object.defineProperty(exports, "R1Storage", { enumerable: true, get: function () { return storage_1.R1Storage; } });
Object.defineProperty(exports, "storage", { enumerable: true, get: function () { return storage_1.storage; } });
// LLM and messaging APIs
var llm_1 = require("./llm");
Object.defineProperty(exports, "R1Messaging", { enumerable: true, get: function () { return llm_1.R1Messaging; } });
Object.defineProperty(exports, "LLMHelpers", { enumerable: true, get: function () { return llm_1.LLMHelpers; } });
Object.defineProperty(exports, "messaging", { enumerable: true, get: function () { return llm_1.messaging; } });
Object.defineProperty(exports, "llmHelpers", { enumerable: true, get: function () { return llm_1.llmHelpers; } });
// UI utilities
var ui_1 = require("./ui");
Object.defineProperty(exports, "CSSUtils", { enumerable: true, get: function () { return ui_1.CSSUtils; } });
Object.defineProperty(exports, "DOMUtils", { enumerable: true, get: function () { return ui_1.DOMUtils; } });
Object.defineProperty(exports, "LayoutUtils", { enumerable: true, get: function () { return ui_1.LayoutUtils; } });
Object.defineProperty(exports, "PerformanceUtils", { enumerable: true, get: function () { return ui_1.PerformanceUtils; } });
Object.defineProperty(exports, "R1Component", { enumerable: true, get: function () { return ui_1.R1Component; } });
Object.defineProperty(exports, "R1_DIMENSIONS", { enumerable: true, get: function () { return ui_1.R1_DIMENSIONS; } });
// New UI design system
Object.defineProperty(exports, "R1UI", { enumerable: true, get: function () { return ui_1.R1UI; } });
Object.defineProperty(exports, "ui", { enumerable: true, get: function () { return ui_1.ui; } });
// Media APIs
var media_1 = require("./media");
Object.defineProperty(exports, "CameraAPI", { enumerable: true, get: function () { return media_1.CameraAPI; } });
Object.defineProperty(exports, "MicrophoneAPI", { enumerable: true, get: function () { return media_1.MicrophoneAPI; } });
Object.defineProperty(exports, "SpeakerAPI", { enumerable: true, get: function () { return media_1.SpeakerAPI; } });
Object.defineProperty(exports, "MediaUtils", { enumerable: true, get: function () { return media_1.MediaUtils; } });
Object.defineProperty(exports, "camera", { enumerable: true, get: function () { return media_1.camera; } });
Object.defineProperty(exports, "microphone", { enumerable: true, get: function () { return media_1.microphone; } });
Object.defineProperty(exports, "speaker", { enumerable: true, get: function () { return media_1.speaker; } });
const hardware_2 = require("./hardware");
const storage_2 = require("./storage");
const llm_2 = require("./llm");
const ui_2 = require("./ui");
const media_2 = require("./media");
/**
 * R1 SDK main class
 * Provides a unified interface to all R1 capabilities
 */
class R1SDK {
    constructor() {
        // Hardware access
        this.accelerometer = hardware_2.accelerometer;
        this.touch = hardware_2.touch;
        this.hardware = hardware_2.hardwareEvents;
        this.deviceControls = hardware_2.deviceControls;
        // Storage
        this.storage = storage_2.storage;
        // LLM and messaging
        this.messaging = llm_2.messaging;
        this.llm = llm_2.llmHelpers;
        // Media
        this.camera = media_2.camera;
        this.microphone = media_2.microphone;
        this.speaker = media_2.speaker;
        // UI
        this.ui = ui_2.ui;
        // Constants
        this.DIMENSIONS = ui_2.R1_DIMENSIONS;
    }
    /**
     * Initialize the SDK and check environment
     */
    async initialize() {
        // Check if running in R1 environment
        if (typeof window === 'undefined') {
            throw new Error('R1 SDK must be used in a browser environment');
        }
        // Log SDK initialization
        console.log('R1 SDK initialized');
        console.log(`Display dimensions: ${ui_2.R1_DIMENSIONS.width}x${ui_2.R1_DIMENSIONS.height}px`);
        // Check available features
        const features = await this.getAvailableFeatures();
        console.log('Available features:', features);
    }
    /**
     * Get available R1 features
     */
    async getAvailableFeatures() {
        return {
            storage: storage_2.R1Storage.isAvailable(),
            secureStorage: storage_2.R1Storage.isSecureAvailable(),
            accelerometer: await this.accelerometer.isAvailable(),
            camera: await this.camera.isAvailable(),
            microphone: await this.microphone.isAvailable(),
            speaker: await media_2.MediaUtils.isSupported('speaker'),
            messaging: typeof PluginMessageHandler !== 'undefined'
        };
    }
    /**
     * Create a simple R1 plugin with common setup
     * @param config Plugin configuration
     */
    createPlugin(config) {
        return new R1Plugin(config);
    }
}
exports.R1SDK = R1SDK;
/**
 * Simple plugin class for quick R1 app development
 */
class R1Plugin {
    constructor(config) {
        this.config = config;
        this.mounted = false;
        this.setupEventListeners();
    }
    mount() {
        if (this.mounted)
            return;
        this.mounted = true;
        this.config.onMount?.();
    }
    unmount() {
        if (!this.mounted)
            return;
        this.mounted = false;
        this.config.onUnmount?.();
    }
    setupEventListeners() {
        // Setup message handling
        if (this.config.onMessage) {
            llm_2.messaging.onMessage(this.config.onMessage);
        }
        // Setup hardware event handling
        if (this.config.onHardwareEvent) {
            const events = [
                'sideClick',
                'longPressStart',
                'longPressEnd',
                'scrollUp',
                'scrollDown'
            ];
            events.forEach(event => {
                hardware_2.hardwareEvents.on(event, () => {
                    this.config.onHardwareEvent?.(event);
                });
            });
        }
    }
}
exports.R1Plugin = R1Plugin;
// Create and export default SDK instance
exports.r1 = new R1SDK();
// Export convenience functions
const initializeR1 = () => exports.r1.initialize();
exports.initializeR1 = initializeR1;
/**
 * Quick setup function for simple R1 plugins
 * @param setup Setup function that receives the SDK instance
 */
const createR1App = (setup) => {
    document.addEventListener('DOMContentLoaded', async () => {
        await exports.r1.initialize();
        setup(exports.r1);
    });
};
exports.createR1App = createR1App;
// Default export
exports.default = exports.r1;
