// r1-create/dist/hardware/device-controls.js
"use strict";
/**
 * Device Controls module for R1 hardware interaction
 * Provides convenient wrappers for scroll wheel and side button events
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.deviceControls = exports.DeviceControls = void 0;
/**
 * Device Controls for R1 hardware (scroll wheel, side button)
 */
class DeviceControls {
    constructor() {
        this.sideButtonEnabled = true;
        this.scrollWheelEnabled = true;
        this.eventListeners = new Map();
    }
    /**
     * Initialize device controls
     * @param options Configuration options
     */
    init(options = {}) {
        this.sideButtonEnabled = options.sideButtonEnabled ?? true;
        this.scrollWheelEnabled = options.scrollWheelEnabled ?? true;
        if (this.sideButtonEnabled) {
            this.setupSideButtonListener();
        }
        if (this.scrollWheelEnabled) {
            this.setupScrollWheelListener();
        }
        // Setup keyboard fallback for development
        if (options.keyboardFallback !== false) {
            this.setupKeyboardFallback();
        }
    }
    /**
     * Setup side button event listener
     */
    setupSideButtonListener() {
        window.addEventListener('sideClick', (event) => {
            if (!this.sideButtonEnabled)
                return;
            this.handleSideButtonClick(event);
        });
    }
    /**
     * Setup scroll wheel event listener
     */
    setupScrollWheelListener() {
        // Flutter sends scrollUp and scrollDown events instead of wheel events
        window.addEventListener('scrollUp', (event) => {
            if (!this.scrollWheelEnabled)
                return;
            this.handleScrollWheel({ direction: 'up', event });
        });
        window.addEventListener('scrollDown', (event) => {
            if (!this.scrollWheelEnabled)
                return;
            this.handleScrollWheel({ direction: 'down', event });
        });
    }
    /**
     * Setup keyboard fallback (space bar = side button)
     */
    setupKeyboardFallback() {
        window.addEventListener('keydown', (event) => {
            if (event.code === 'Space') {
                event.preventDefault();
                const sideClickEvent = new CustomEvent('sideClick', {
                    detail: { source: 'keyboard' }
                });
                window.dispatchEvent(sideClickEvent);
            }
        });
    }
    /**
     * Handle side button click
     * @param event The side click event
     */
    handleSideButtonClick(event) {
        const handlers = this.eventListeners.get('sideButton') || new Set();
        handlers.forEach(handler => handler(event));
    }
    /**
     * Handle scroll wheel events
     * @param data The scroll data
     */
    handleScrollWheel(data) {
        const handlers = this.eventListeners.get('scrollWheel') || new Set();
        handlers.forEach(handler => handler(data));
    }
    /**
     * Register event handler
     * @param eventType 'sideButton' or 'scrollWheel'
     * @param handler Event handler function
     */
    on(eventType, handler) {
        if (!this.eventListeners.has(eventType)) {
            this.eventListeners.set(eventType, new Set());
        }
        this.eventListeners.get(eventType).add(handler);
    }
    /**
     * Remove event handler
     * @param eventType 'sideButton' or 'scrollWheel'
     * @param handler Event handler function to remove
     */
    off(eventType, handler) {
        const handlers = this.eventListeners.get(eventType);
        if (handlers) {
            handlers.delete(handler);
        }
    }
    /**
     * Enable or disable side button
     * @param enabled Enable state
     */
    setSideButtonEnabled(enabled) {
        this.sideButtonEnabled = enabled;
    }
    /**
     * Enable or disable scroll wheel
     * @param enabled Enable state
     */
    setScrollWheelEnabled(enabled) {
        this.scrollWheelEnabled = enabled;
    }
    /**
     * Simulate side button click (for testing)
     */
    triggerSideButton() {
        const event = new CustomEvent('sideClick', {
            detail: { source: 'programmatic' }
        });
        window.dispatchEvent(event);
    }
    /**
     * Check if side button is enabled
     */
    isSideButtonEnabled() {
        return this.sideButtonEnabled;
    }
    /**
     * Check if scroll wheel is enabled
     */
    isScrollWheelEnabled() {
        return this.scrollWheelEnabled;
    }
    /**
     * Get all registered event listeners for debugging
     */
    getEventListeners() {
        const result = {};
        this.eventListeners.forEach((handlers, eventType) => {
            result[eventType] = handlers.size;
        });
        return result;
    }
}
exports.DeviceControls = DeviceControls;
// Export singleton instance
exports.deviceControls = new DeviceControls();
