// r1-create/dist/hardware/index.js
"use strict";
/**
 * Hardware API module for R1 device access
 * Provides accelerometer, touch simulation, and hardware button events
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.deviceControls = exports.DeviceControls = exports.hardwareEvents = exports.touch = exports.accelerometer = exports.HardwareEvents = exports.TouchAPI = exports.AccelerometerAPI = void 0;
class AccelerometerAPI {
    constructor() {
        this.isListening = false;
    }
    /**
     * Check if accelerometer is available on the device
     */
    async isAvailable() {
        return await window.creationSensors?.accelerometer?.isAvailable() ?? false;
    }
    /**
     * Start receiving accelerometer data
     * @param callback Function to handle accelerometer data
     * @param options Configuration options including frequency
     */
    start(callback, options) {
        if (this.isListening) {
            this.stop();
        }
        this.currentCallback = callback;
        this.isListening = true;
        if (window.creationSensors?.accelerometer) {
            window.creationSensors.accelerometer.start(callback, options);
        }
    }
    /**
     * Stop receiving accelerometer data
     */
    stop() {
        if (this.isListening && window.creationSensors?.accelerometer) {
            window.creationSensors.accelerometer.stop();
            this.isListening = false;
            this.currentCallback = undefined;
        }
    }
    /**
     * Check if currently listening for accelerometer data
     */
    isActive() {
        return this.isListening;
    }
}
exports.AccelerometerAPI = AccelerometerAPI;
class TouchAPI {
    /**
     * Simulate a tap at specific coordinates
     * @param x X coordinate (0-240)
     * @param y Y coordinate (0-282)
     */
    tap(x, y) {
        this.sendTouchEvent({ type: 'tap', x, y });
    }
    /**
     * Simulate touch down event
     * @param x X coordinate
     * @param y Y coordinate
     */
    touchDown(x, y) {
        this.sendTouchEvent({ type: 'down', x, y });
    }
    /**
     * Simulate touch up event
     * @param x X coordinate
     * @param y Y coordinate
     */
    touchUp(x, y) {
        this.sendTouchEvent({ type: 'up', x, y });
    }
    /**
     * Simulate touch move event
     * @param x X coordinate
     * @param y Y coordinate
     */
    touchMove(x, y) {
        this.sendTouchEvent({ type: 'move', x, y });
    }
    /**
     * Cancel touch event
     * @param x X coordinate
     * @param y Y coordinate
     */
    touchCancel(x, y) {
        this.sendTouchEvent({ type: 'cancel', x, y });
    }
    sendTouchEvent(event) {
        if (typeof TouchEventHandler !== 'undefined') {
            TouchEventHandler.postMessage(JSON.stringify(event));
        }
    }
}
exports.TouchAPI = TouchAPI;
class HardwareEvents {
    constructor() {
        this.listeners = new Map();
        this.initializeEventListeners();
    }
    /**
     * Add event listener for hardware button events
     * @param event Event type to listen for
     * @param callback Function to call when event occurs
     */
    on(event, callback) {
        if (!this.listeners.has(event)) {
            this.listeners.set(event, new Set());
        }
        this.listeners.get(event).add(callback);
    }
    /**
     * Remove event listener
     * @param event Event type
     * @param callback Function to remove
     */
    off(event, callback) {
        const callbacks = this.listeners.get(event);
        if (callbacks) {
            callbacks.delete(callback);
        }
    }
    /**
     * Remove all listeners for an event type
     * @param event Event type to clear
     */
    removeAllListeners(event) {
        if (event) {
            this.listeners.delete(event);
        }
        else {
            this.listeners.clear();
        }
    }
    initializeEventListeners() {
        // Only initialize in browser environment
        if (typeof window === 'undefined')
            return;
        // Side button (PTT) events
        window.addEventListener('sideClick', () => {
            this.emit('sideClick');
        });
        window.addEventListener('longPressStart', () => {
            this.emit('longPressStart');
        });
        window.addEventListener('longPressEnd', () => {
            this.emit('longPressEnd');
        });
        // Scroll wheel events
        window.addEventListener('scrollUp', () => {
            this.emit('scrollUp');
        });
        window.addEventListener('scrollDown', () => {
            this.emit('scrollDown');
        });
    }
    emit(event) {
        const callbacks = this.listeners.get(event);
        if (callbacks) {
            callbacks.forEach(callback => callback());
        }
    }
}
exports.HardwareEvents = HardwareEvents;
// Export singleton instances
exports.accelerometer = new AccelerometerAPI();
exports.touch = new TouchAPI();
exports.hardwareEvents = new HardwareEvents();
// Re-export device controls
var device_controls_1 = require("./device-controls");
Object.defineProperty(exports, "DeviceControls", { enumerable: true, get: function () { return device_controls_1.DeviceControls; } });
Object.defineProperty(exports, "deviceControls", { enumerable: true, get: function () { return device_controls_1.deviceControls; } });
